File: Dustbin.py
"""
Dustbin Class

Contains the main switchboard for the project, connecting all other components.

Holds global variables and global methods.
"""

from communication.Communicator import Communicator
from memory.ShortTermMemory import ShortTermMemory
from robot.Robot import Robot
from time import time
from time import sleep
from communication.Switchboard import Switchboard
from communication.Events import Events
from machine_learning.Vision import Vision
from threading import Thread

import socket
def internet(host="8.8.8.8", port=53, timeout=3):
    """
    Host: 8.8.8.8 (google-public-dns-a.google.com)
    OpenPort: 53/tcp
    Service: domain (DNS/TCP)
    """
    try:
        socket.setdefaulttimeout(timeout)
        socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
        return True
    except Exception as ex:
        print ex.message
        return False

class ShutdownListener(Events.EventListener) :
    def callback(self, kwargs) :
        return self.container.done()

class Dustbin :
    def __init__(self, logger, audio_timeout, silent) :
        try :
            self.logger = logger
            self.log = self.logger.log
            self.initVars(silent)
            # First need a switchboard for subscription
            self.switchboard = Switchboard(self)
            self.switchboardThread = Thread(target = self.switchboard.run)
            self.switchboardThread.start()
            # Then need a communicator
            self.com = Communicator(audio_timeout, self)
            # Then short term memory
            self.memory = ShortTermMemory(self)
            # Finally can add a robot
            self.robot = Robot(self)
            self.robotThread = Thread(target = self.robot.run)
            self.robotThread.start()
            self.vision = Vision(self)
            self.visionThread = Thread(target = self.vision.run)
            self.visionThread.start()
            # End by subscribing to shutdown
            self.shutdownListener = ShutdownListener(self)
            self.subscribe(Events.REQ_SHUTDOWN, self.shutdownListener)
        except Exception as e :
            print 'FATAL ERROR OCCURRED DURING SETUP!' , e
            self.done()
            exit(1)
    def getLogLock(self) :
        return self.logger.logLock
    def initVars(self, silent) :
        self.keepGoing = True
        self._REFRESH_RATE = 60
        self.silent = silent
        # Checks for internet connection after this amount of time.
        self._logTime = time()
        self._hasInternet = internet()
    def subscribe(self, event, listener, callback=None) :
        self.switchboard.subscribe(event, listener, callback)
    def trigger(self, event, **kwargs) :
        kwargs['event'] = event
        self.switchboard.runTrigger(event, kwargs)
    def done(self) :
        self.keepGoing = False
        # NEVER REMOVE THIS LINE! Somehow, it is necessary.
        if hasattr(self, 'robot') and self.robot is not None:
            self.robot.end()
        if hasattr(self, 'vision') and self.vision is not None :
            self.vision.stop()
        if hasattr(self, 'switchboard') and self.switchboard is not None :
            self.switchboard.stop()
        if hasattr(self, 'robotThread') and self.robotThread is not None :
            self.robotThread.join()
        if hasattr(self, 'visionThread') and self.visionThread is not None :
            self.visionThread.join()
        if hasattr(self, 'switchboardThread') and self.switchboardThread is not None :
            self.switchboardThread.join()
    def runCommands(self, commands, callback) :
        self.callback = callback
        for command in commands :
            if '.wav' in command:
                self.com.interpretFromWavFile(command)
            else :
                self.com.interpretText(command)
        sleep(2) # Hokey way to wait for processes to finish
        if self.callback is not None :
            self.callback()
    def run(self) :
        try :
            while self.keepGoing :
                response = self.com.interpretAudio()
                sleep(0.5)
                self.log('Listening process continues.')
            self.done()
        except Exception as e:
            self.log('Dustbin process had fatal error.', e)
            print('Dustbin process had fatal error.', e)
            self.done()
            exit(1)
    def hasInternet(self) :
        # After REFRESH_RATE seconds, should check again for internet connection
        if (time() - self._logTime) > self._REFRESH_RATE :
            self._hasInternet = internet()
            self._logTime = time()
        return self._hasInternet

File: Dustbin.js
'use strict';

const { Events, EventListener, MyEventEmitter } = require('./communication/Events.js');
const AudioCommunicator = require('./communication/AudioCommunicator.js').AudioCommunicator;
const ShortTermMemory = require('./memory/ShortTermMemory.js').ShortTermMemory;
const Robot = require(`${__dirname}/robot/Robot.js`);
const Vision = require(`${__dirname}/machine_learning/Vision.js`).Vision;

/**
 * Dustbin Class
 * Dustbin is the main wrapper class for the robot. It coordinates the vision, memory, speech, and movement components.
 * Holds global variables and global methods.
 * @module Dustbin
 * @namespace Dustbin
 * @class Dustbin
 * @classdesc This is a description of the MyClass class.
 */
class Dustbin {
  /**
   * 
   * @param {*} _logger Logger class instance.
   * @param {*} audio_timeout Time waiting for audio before giving up and trying to interpret.
   * @param {*} silent If silent is true, then will not play audio. Otherwise, will speak back responses.
   */
  constructor(_logger, audio_timeout, silent, _testing=false) {
    this.testing = _testing;
    this._lastCheck = Date.now();
    this.logger = _logger;
    this.log('-------------------DUSTBIN BEGINS-------------------');
    this.keepGoing = true;
    this.ended = false;
    this._REFRESH_RATE = 60;
    this.silent = silent;
    this.switchboard = new MyEventEmitter();
    try {
      // # Then need a communicator
      this.com = new AudioCommunicator(audio_timeout, this);
      // # Then short term memory
      this.memory = new ShortTermMemory(this);
      // # Finally can add a robot
      this.robot = new Robot(this);
      // this.robotThread = Thread(target = this.robot.run);
      // this.robotThread.start();
      this.vision = new Vision(this);
      // this.visionThread = Thread(target = this.vision.run);
      // this.visionThread.start();
      // # End by subscribing to shutdown
      this.subscribe(new EventListener(Events.REQ_SHUTDOWN, (kwargs)=>this.done(kwargs)));
    } catch (e) {
      console.log('FATAL ERROR OCCURRED DURING SETUP!', e);
      this.done();
    }
  }
  /**
   * Tests for an internet connection.
   * Returns a Promise which resolves/rejects boolean true/false.
   */
  _testInternetPromise() {
    return new Promise(function (resolve, reject) {
      require('dns').resolve('www.google.com', function (err) {
        if (err) {
          reject(false);
        } else {
          resolve(true);
        }
      });
    })
  }
  /**
   * Calls the Logger.log function. See Logger.log
   */
  log() {
    this.logger.log(arguments);
  }
  /**
   * Subscribes an EventListener
   * @param {*} listener Listener object, includes callback for the triggered event.
   */
  subscribe(listener) {
    return this.switchboard.subscribe(listener);
  }
  
  /**
   * Returns a Promise.
   * This tells the Dustbin that it's time to stop.
   */
  done(testing=false) {
    if (this.ended) { return; }
    this.keepGoing = false;
    const self = this;
    if (self.robot)
      self.robot.end();
    self.log('-------------------DUSTBIN   ENDS-------------------');
    self.logger.end();
    self.ended = true;
    if (!testing) {
      process.exit(0);
    }
    return Promise.resolve();
  }
  /**
   * Tests whether or not internet is connected, as long as it has been enough time since last checked.
   * Updates this._hasInternet flag accordingly.
   */
  testInternet() {
    // After REFRESH_RATE seconds, should check again for internet connection
    if ((Date.now() - this._lastCheck) > this._REFRESH_RATE) {
      let self = this;
      this._testInternetPromise().then(i => { self._hasInternet = i });
      this._lastCheck = Date.now();
    }
    return this._hasInternet;
  }
  /**
   * Persistent Dustbin run loop. Checks for internet if is not connected. Interprets audio, then handles response.
   */
  run() {
    try {
      while (this.keepGoing) {
        if (!this._hasInternet) {
          this.testInternet();
        }
        setTimeout(() => {
          this.com.interpretAudio().then(response => {
            // TODO is this even necessary, or do we handle it in interpret audio instead?
          })
          this.log('Listening process continues.');
          // TODO is the timeout necessary if we are already waiting while getting audio?
        }, 500);
      }
      this.done();
    }
    catch (e) {
      this.log('Dustbin process had fatal error.', e);
      console.log('Dustbin process had fatal error.', e);
      return this.done();
    }
  }
  /**
   * Returns MyEventEmitter.emit (which is a Promise.all for all listeners triggered)
   * @param {*} event Integer for the event to trigger.
   * @param {*} kwargs Arguments to pass to all listeners for this event.
   */
  trigger(event, kwargs = {}) {
    kwargs.event = event;
    return this.switchboard.emit(event, kwargs);
  }
  /**
   * Rather than running persistently, the dustbin will run a list of commands in sequence. This is primarily for testing purposes.
   * @param {*} commands List of commands to interpret. Each command is interpretted separately.
   */
  runCommands(commands) {
    let promise = Promise.resolve();
    for (const command of commands) {
      if (command.indexOf('.wav') != -1) {
        promise = promise.then(() => { return this.com.interpretFromWavFile(command); });
      }
      else {
        promise = promise.then(() => { return this.com.interpretText(command); });
      }
    }
    return promise;
  }
}

module.exports = Dustbin;